### PR TITLE
Simon/send tx

### PR DESCRIPTION
**Pull request by sdschmidt**
add input field for send transaction form.

### CODE REVIEW

**Comment by danimoh** [comment]
Which one is the parent branch of your work?
The `soeren/merged` branch is not gonna be merged into master.
Please create a pull request that only contains your changes, e.g. cherry pick your commits onto a branch that's based on master.

File: src/components/SendTx.vue [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        width: 100%;
+        max-width: calc(56.25 * var(--nimiq-size, 8px)); /* 450px */
+        min-height: calc(58 * var(--nimiq-size, 8px)); /* 464px */
+        max-height: calc(77.5 * var(--nimiq-size, 8px)); /* 620px */
+        display: flex;
+        flex-direction: column;
+        flex-shrink: 0;
+        margin: 0 calc(1.25 * var(--nimiq-size, 8px));
+    }
+</style>
```
Comment (danimoh) [inline src/components/SendTx.vue]: The style in here should be ordered similar to their appearance in the HTML

File: src/components/SendTx.vue [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @media (max-width: 450px) {
+        .nq-card {
+            margin: calc(2 * var(--nimiq-size, 8px)) calc(1 * var(--nimiq-size, 8px));
+        }
+    }
+    /** END Nimiq Style **/
+
+    .identicon-Contact {
```
Comment (danimoh) [inline src/components/SendTx.vue]: I think this style is not applied due to style scoping.

File: src/components/SendTx.vue [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        margin-left: 15px !important;
+    }
+
+    .small-page {
```
Comment (danimoh) [inline src/components/SendTx.vue]: SmallPage not part of this component

File: src/main.ts [hunk 4]
```diff
@@ -8,6 +8,9 @@ export { default as AmountWithDetails } from './components/AmountWithDetails.vue
 export { default as Contact } from './components/Contact.vue';
 export { default as ContactList } from './components/ContactList.vue';
 export { default as Identicon } from './components/Identicon.vue';
+export { default as IdenticonContact } from './components/IdenticonContact.vue';
+export { default as InputAddress } from './components/InputAddress.vue';
+export { default as InputTextChunk } from './components/InputTextChunk.vue';
```
Comment (danimoh) [inline src/main.ts]: `InputTextChunk` unused

File: src/stories/index.stories.js [hunk 5]
```diff
@@ -65,7 +68,40 @@ storiesOf('Basic', module)
...
+        const name = 'Nimiq';
+        return {
+            components: {IdenticonContact},
+            data() {
+                return { address, name };
+            },
+            template: `<IdenticonContact :address="address" :name="name"/>`,
+        };
+    })    
+    .add('IdenticonContact (empty)', () => {
```
Comment (danimoh) [inline src/stories/index.stories.js]: I don't think it's necessary to create a story just for that.
A knob for the address is defined anyways.
Might want to define one for the name also.

File: src/stories/index.stories.js [hunk 6]
```diff
@@ -65,7 +68,40 @@ storiesOf('Basic', module)
...
+        const name = '';
+        return {
+            components: {IdenticonContact},
+            data() {
+                return { address, name };
+            },
+            template: `<IdenticonContact :address="address" :name="name"/>`,
+        };
+    })
+    .add('InputAddress (empty)', () => {
```
Comment (danimoh) [inline src/stories/index.stories.js]: remove the ` (empty)`

File: src/stories/index.stories.js [hunk 7]
```diff
@@ -65,7 +68,40 @@ storiesOf('Basic', module)
...
+        const address = text('address', '');
+        const name = '';
+        return {
+            components: {InputAddress},
+            data() {
+                return { address, name };
+            },
+            template: `<InputAddress :address="address" :name="name"/>`,
```
Comment (danimoh) [inline src/stories/index.stories.js]: Does not have a prop `name`

File: src/stories/index.stories.js [hunk 8]
```diff
@@ -412,6 +448,48 @@ storiesOf('Components', module)
...
+                <small-page style="height: 560px;">
+                    <SendTx address="NQ72 YBV0 9SSA XBAN 820M 8H0B 339V FJP4 34DH" :disabled="true">
+                        <IdenticonContact name="Simon" address="NQ72 YBV0 9SSA XBAN 820M 8H0B 339V FJP4 34DH"/>
+                        <IdenticonContact name="Name" address="NQ79 XYXY FE3L 7CQL KAN3 KRVK 2P0L B33D JQ81"/>
+                        <IdenticonContact name="" address=""/>
+                    </SendTx>
+                </small-page>
+            `)
+        };
+    })
```
Comment (danimoh) [inline src/stories/index.stories.js]: These should be moved to the pages stories

File: src/stories/index.stories.js [hunk 9]
```diff
@@ -412,6 +448,48 @@ storiesOf('Components', module)
             template: `<div style="width: 400px"><PaymentInfoLine :amount="${amount}" :fee="${fee}" origin="${origin}"/></div>`,
         };
     })
+    .add('SendTx', () => {
+        return {
+            components: {SendTx, SmallPage, IdenticonContact},
+            template: windowTemplate(`
+                <small-page style="height: 560px;">
+                    <SendTx address="NQ10 7" :disabled="false">
```
Comment (danimoh) [inline src/stories/index.stories.js]: false is the default for `disabled` already.

Also I think, SendTx should only accept valid addresses.

Might also want to make the address as knob.

File: src/stories/index.stories.js [hunk 10]
```diff
@@ -412,6 +448,48 @@ storiesOf('Components', module)
...
+                    <SendTx address="">
+                        <IdenticonContact/>
+                        <IdenticonContact/>
+                        <IdenticonContact/>
+                    </SendTx>
+                </small-page>
+            `)
+        };
+    })
+    .add('SendTx (full/disabled)', () => {
```
Comment (danimoh) [inline src/stories/index.stories.js]: Make address and disabled knobs and remove this story

File: src/stories/index.stories.js [hunk 11]
```diff
@@ -412,6 +448,48 @@ storiesOf('Components', module)
...
+                    <SendTx address="NQ10 7" :disabled="false">
+                        <IdenticonContact name="Simon" address="NQ72 YBV0 9SSA XBAN 820M 8H0B 339V FJP4 34DH"/>
+                        <IdenticonContact name="Name" address="NQ79 XYXY FE3L 7CQL KAN3 KRVK 2P0L B33D JQ81"/>
+                        <IdenticonContact name="" address=""/>
+                    </SendTx>
+                </small-page>
+            `)
+        };
+    })
+    .add('SendTx (empty)', () => {
```
Comment (danimoh) [inline src/stories/index.stories.js]: might also want to make the contacts knobs and then remove this story

File: src/stories/index.stories.js [hunk 12]
```diff
@@ -433,8 +511,10 @@ storiesOf('Components', module)
             template: windowTemplate(`<OnboardingMenu @signup="signup" @login="login" @ledger="ledger"/>`),
         };
     });
-
 storiesOf('Pages', module)
+storiesOf('Pages/SendTx', module)
+    .addDecorator(withKnobs)
```
Comment (danimoh) [inline src/stories/index.stories.js]: Empty category

File: src/components/IdenticonContact.vue [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+<style scoped>
+
+    span {
+        text-align: center;
+    }
+
+    .identicon-contact {
+        width: calc(11 * var(--nimiq-size, 8px));
+        min-width: calc(5 * var(--nimiq-size, 8px));
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/IdenticonContact.vue]: removed line 29 `min-width: calc(5 * var(--nimiq-size, 8px));`

File: src/components/IdenticonContact.vue [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    @Prop(String) public address!: string;
+    @Prop(String) public name!: string;
+}
+</script>
+
+<style scoped>
+
+    span {
+        text-align: center;
+    }
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/IdenticonContact.vue]: removed lines 23-25

File: src/components/IdenticonContact.vue [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+    .identicon-contact {
+        width: calc(11 * var(--nimiq-size, 8px));
+        min-width: calc(5 * var(--nimiq-size, 8px));
+        text-align: center;
+        margin: 0;
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/IdenticonContact.vue]: remove line 31

File: src/components/IdenticonContact.vue [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        width: 72%;
+    }
+
+    .name-placeholder {
+        background-color: var(--nimiq-gray);
+        width: 72%;
+        height: 0.5em;
+        margin: auto;
+        margin-top: 8px;
+        border-radius: 0.5em;
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/IdenticonContact.vue]: replaced on lines 42,44,45

File: src/components/IdenticonContact.vue [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    img {
+        width: 100%;
+        height: 100%;
+    }
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/IdenticonContact.vue]: removed lines 48-51

File: src/components/IdenticonContact.vue [hunk 18]
```diff
@@ -0,0 +1,52 @@
+<template>
+    <div class="identicon-contact">
+        <Identicon :address="address" />
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/IdenticonContact.vue]: the new plain grey placeholder will be used on signup and ledger in Keyguard & download login file. Id suggest keeping it consistent and therefore changing it out in the /src/components/Identicon.vue

File: src/components/InputAddress.vue [hunk 19]
```diff
@@ -0,0 +1,479 @@
...
+import {Component, Prop, Watch, Vue, Provide} from 'vue-property-decorator';
+import '@nimiq/style/nimiq-style.min.css'
+import InputTextChunk from './InputTextChunk.vue'
+
+function preFormatAddress(val: string) {
+    let plain:string
+    try { 
+        // remove invalid characters
+        plain = val.trim()
+                   .replace(/[wizo]/gi,'')
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: cool idea changed that

File: src/components/InputAddress.vue [hunk 20]
```diff
@@ -0,0 +1,479 @@
...
+                   .replace(/[\W](?!$)/gi,'')
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: /[\W](?!$)/ doesn't match trailing spaces

File: src/components/InputAddress.vue [hunk 21]
```diff
@@ -0,0 +1,479 @@
...
+            }
+        } else if (plain.length == 2) { 
+            // add nq and add all not [nq] characters
+            return 'NQ'+plain.replace(/[nq]/gi,'')
+        } else if (plain[0].match(/\d/)) { 
+            // add NQ to string starting w/ digits
+            return 'NQ'+plain
+        } else { 
+            // replace word characts at begining /w NQ
+            return plain.replace(/^([\D]{1,2})/,"NQ")
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: it replaces word characters when eg. pasting with NQ.

File: src/components/InputAddress.vue [hunk 22]
```diff
@@ -0,0 +1,479 @@
...
+
+function formatAddress(val: any):string {
+    if (val !instanceof String) {return}    
+    const trailing:boolean = (val[val.length - 1] == ' ')
+    const pre = preFormatAddress(val)
+    if (pre.length == 0) {return ''}
+    const tokens:string[] = pre.match(/\w{1,4}/g) 
+    let address:string = tokens.join(' ')
+    if ((trailing) && (address.length%5 == 4)) {
+        address += ' '
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: pre.match(/\w{1,4}/g) divides up the address into chunks of max. 4 characters. This removes all non-word characters. If there was one before (`trailing`) then we add it again on line 57.

File: src/components/InputAddress.vue [hunk 23]
```diff
@@ -0,0 +1,479 @@
...
+    }
+    address = address.split('')
+        try { // insert linebreak
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: removed linebreaks

File: src/components/InputAddress.vue [hunk 24]
```diff
@@ -0,0 +1,479 @@
...
+
+function preFormatAddress(val: string) {
+    let plain:string
+    try { 
+        // remove invalid characters
+        plain = val.trim()
+                   .replace(/[wizo]/gi,'')
+                   .replace(/[\W](?!$)/gi,'')
+                   .toUpperCase()
+    } catch(err) {constole.log(err); return ''}
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: removed the catch block

File: src/components/InputAddress.vue [hunk 25]
```diff
@@ -0,0 +1,479 @@
...
+  document.body.appendChild(textArea);
+  textArea.focus();
+  textArea.select();
+
+  try {
+    var successful = document.execCommand('copy');
+    var msg = successful ? 'successful' : 'unsuccessful';
+  } catch (err) { console.log(err )}
+
+  document.body.removeChild(textArea);
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: remove function copyTextToClipboard

File: src/components/InputAddress.vue [hunk 26]
```diff
@@ -0,0 +1,479 @@
...
+            let k:string = this.preCalcTextInput(event.key)      
+            if ((k.c+1)%5 == 0) {
+                // add space when at end of box
+                k = this.preCalcTextInput(event.key+' ')
+            }
+            t.value = k.n
+            t.selectionStart = k.c 
+            t.selectionEnd = k.c 
+            } else {
+                // trigger bad char event here
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: I thought we could add an animation here, but replacing WIZO with V120 respectively is the better option

File: src/components/InputAddress.vue [hunk 27]
```diff
@@ -0,0 +1,479 @@
...
+            const l:integer = (Math.floor(t.value.length/5)+1)*5
+
+            
+            if (event.shiftKey) {
+                if ((t.selectionStart == t.selectionEnd) && (t.selectionEnd == 0)) {
+                    return
+                }
+                event.preventDefault()
+                t.selectionEnd = (ce-10 > 0) ? ce-6 : 0
+                t.selectionStart = (ce-10 > 0) ? ce-10: 0
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: when eg entering 'NQ07 0000 0000' reaching the very first block with Shift+Tab, it overflows to the last block.

File: src/components/InputAddress.vue (hunk 23 above)
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: The line breaks are actually needed: 
screenshot 2019-02-11 at 17 25 39

File: src/components/InputAddress.vue [hunk 28]
```diff
@@ -0,0 +1,479 @@
...
+        const specialKey:boolean = (event.altKey) || (event.ctrlKey) || (event.metaKey)
+
+        if (specialKey) {
+            this.updateDisplay()
+            return
+        }
+
+        if (event.key == 'Enter') {event.preventDefault()}
+
+        this.lockInput(t)
```
[the remark this answers is not in the document]
Comment (sdschmidt) [reply src/components/InputAddress.vue]: it prevented jiggly behaviour in an earlier version. removed.

**Comment by danimoh** [comment]
Closed in favor of https://github.com/nimiq/vue-components/pull/43 and https://github.com/nimiq/vue-components/tree/sebastian/send-tx.
Some of your ideas have been reused, thank you.

**Pull request closed without merging**